### PR TITLE
feat(*): use extra options to get redirect url on pc_pay & h5_pay

### PR DESCRIPTION
**Pull request by ZhenhangTung**
I made breaking changes on these two API. The reason is that the current way of returning HTML code is not reasonable for the frontend, so I checked your code and you basically generate a form and request the JD's API using `POST` method.
I monitored all the requests and responses of JDPay and found that there were 302 redirects.
What we can do is to get the URL that being redirected to and return it to the frontend, so the browser could do a redirect, instead of trying to render HTML code. 
Furthermore, I renamed some functions to fix grammar mistakes. The deprecated API `notify_verify` is still gonna work for backward compatibility has been made.
Please comment on this PR if you have any thoughts.

### CODE REVIEW

**Comment by ZhenhangTung** [comment]
Oh, I missed that I've also added a new API to verify redirect after payment being finished.
Thanks for your work on this project.

**Comment by genkin-he** [comment]
Thanks for your work, I truly want to merge your pull request, but this gem is working in production in our company。Breaking changes is not allowed at current version, So perhaps you should add a new method or optional parameter to get the URL.

**Comment by ZhenhangTung** [comment]
If that's the case, I could just add some new methods and keep the old ones, and you can upgrade to thoese methods when you're ready.

**Comment by ZhenhangTung** [comment]
Hey @genkin-he , I created a new module called `V1`, and keep the API's name is the same as the old ones. 
I believe this approach requires minium changes and you guys could move current API into `V1` and freeze them when you're ready to bump this project to `V1.0`.

**Comment by ZhenhangTung** [comment]
Hey @genkin-he , any update on this PR?

**Comment by genkin-he** [comment]
> Breaking changes is not allowed

I think you misunderstood me on this.

**Comment by ZhenhangTung** [comment]
Nope, I believe I'm understanding it correctly, so that's why I've taken your advice and revert all my changes in the current service module and create a new module called [V1](https://github.com/udacity/jdpay/blob/a6e28011a94115912c37a18c435efeb080f95754/lib/jd_pay/v1/service.rb). I shipped my changes into this module, with same name as old ones, so there wound't be any side effect on your production env.

**Comment by ZhenhangTung** [comment]
As I mentioned before:
> I believe this approach requires minium changes and you guys could move current API into V1 and freeze them when you're ready to bump this project to V1.0.

There will be a time that this project being tagged with V1, with stable API provided, so that's why I offer this solution. If someone want to use new API, all they have to do is add `::V1` to their code and that's it. I've written the migration steps into readme file.

**Comment by genkin-he** [comment]
add a V1 module is a terrible way. you just need to add the `get_redirect_url` method and update the readme, why not? other changes are redundant, it can be left out. 

example:
```ruby
# options
#  need_url, Boolean, default: false
def pc_pay(params, options)
...
options[:need_url] ? get_redirect_url : build_pay_form
end
```
Thanks again! you can add my wechat for discussing it.
wechat number:
```ruby
aa = [0, 1, 3, 6, 7, 8, 9]
bb = [1, 4, 0, 5, 3, 5, 6, 5, 2, 5, 6]
bb.collect { |n| aa[n] }.join
```

**Comment by ZhenhangTung** [comment]
Just explored some projects written in Ruby, it did really rared to see naming package using the version's name, but it's really common and recommended in [this official blog post](https://blog.golang.org/versioning-proposal).
> If we adopt semantic versioning for Go packages, as most Go developers expect, then the import compatibility rule requires that different major versions must use different import paths. This observation led us to semantic import versioning, in which versions starting at v2.0.0 include the major version in the import path: my/thing/v2/sub/pkg.

Then I guess I'll have to comply with Ruby community's consensus.
Just finish the job in the latest commit

**Comment by genkin-he** [comment]
bro, could you squash your commits together into one commit?

**Comment by ZhenhangTung** [comment]
There is an option called `squash and merge` when merging the PR

File: README.md [hunk 1]
```diff
@@ -234,8 +252,31 @@ JdPay::Service.notify_verify(xml_str, options = {})
   </jdpay>
   EOF
 
-  JdPay::Service.notify_verify(notyfy_xml)
+  JdPay::Service.notify_verify(xml_str)
+```
+
```
Comment (genkin-he) [inline README.md]: delete redundant columns

File: README.md [hunk 2]
```diff
@@ -128,7 +142,11 @@ h5_pay
 #### Definition
 
 ```ruby
+# 返回构建的表单，表单会自动发送POST请求，请求成功京东会进行页面跳转
 JdPay::Service.h5_pay(params, options = {})
+
+# 获取Redirect URL，前端获取后可以直接跳转京东收银台
+JdPay::Service.h5_pay(params, {:need_redirect_url => true})
```
Comment (genkin-he) [inline README.md:149]:
```ruby
# options
#   need_redirect_url, Boolean, true: return a url, false: return a html, default: false
JdPay::Service.h5_pay(params, options = {})
```

File: README.md [hunk 3]
```diff
@@ -234,8 +252,31 @@ JdPay::Service.notify_verify(xml_str, options = {})
...
+JdPay::Service.verify_redirection(params, options = {})
+```
+#### Example
+```ruby
+# In your controller:
+begin
+  decrypted_params = JdPay::Service.verify_redirection(request.params, {})
+rescue => e
+  render :status => :bad_request
+  return
```
Comment (genkin-he) [inline README.md:276]:
```ruby
render status: :bad_request
```
1. [Use the Ruby 1.9 hash literal syntax when your hash keys are symbols](https://github.com/rubocop-hq/ruby-style-guide#hash-literals)

2. `return` is redundant.

File: README.md [hunk 4]
```diff
@@ -336,6 +377,8 @@ JdPay::QrService.notify_verify(params, options = {})
 ```ruby
 # refer to：JdPay::Service.notify_verify(params, options = {})
 ```
+
+
```
Comment (genkin-he) [inline README.md]: delete redundant columns

File: lib/jd_pay/error.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+module JdPay
+  module Error
+    class InvalidRedirection < StandardError; end
+  end
+end
```
Comment (genkin-he) [inline lib/jd_pay/error.rb]: add a new line

File: lib/jd_pay/result.rb [hunk 6]
```diff
@@ -16,7 +16,7 @@ def success?
       self['jdpay']['result']['code'] == SUCCESS_FLAG
     end
 
-    def decrypt_verify(options = {})
+    def verify_decrypt(options = {})
```
Comment (genkin-he) [inline lib/jd_pay/result.rb:19]: ? first decrypt, then verify

File: lib/jd_pay/service.rb [hunk 7]
```diff
@@ -39,7 +41,8 @@ def web_pay(params, url, options = {})
           params[k] = skip_encrypt_params.include?(k) ? v : JdPay::Des.encrypt_3des(v)
         end
         params[:sign] = sign
-        JdPay::Util.build_pay_form(url, params)
+        options[:need_redirect_url] ? get_redirect_url(url, params) : JdPay::Util.build_pay_form(url, params)
+
```
Comment (genkin-he) [inline lib/jd_pay/service.rb]: delete redundant columns

File: lib/jd_pay/util.rb [hunk 8]
```diff
@@ -46,6 +46,34 @@ def build_pay_form(url, form_attributes)
...
+            decrypted[k] = v
+            # According to the doc, any empty value should be dropped.
+            next if v == ''
+            new_hash[k] = v
+          end
+        end
+        return decrypted, new_hash
+      end
+
+      def params_to_string(params)
```
Comment (genkin-he) [inline lib/jd_pay/util.rb]: Reinventing the wheel. rename the `to_uri` ?
[Hash#to_query](https://github.com/rails/rails/blob/375a4143cf5caeb6159b338be824903edfd62836/activesupport/lib/active_support/core_ext/object/to_query.rb#L76)

File: README.md [hunk 9]
```diff
@@ -44,6 +44,16 @@ JdPay.qr_pri_key = 'YOUR_RSA_PRIVATE_KEY'
...
 
+*重要*：
+
+根据文档中的Q&A
+> 支付请求和异步返回的加密规则是什么？
+  答：支付请求用pkcs8的RSA私钥进行签名，des加密；
+    同步返回使用京东支付统一对外rsa公钥验签；
+    异步返回用3DES解密，京东支付统一对外rsa公钥验签。
+
+因此请勿为`JdPay.pub_key`设置你自己产生的公钥，默认不填就行。
```
Comment (genkin-he) [inline README.md:55]:
```ruby
# JdPay.pub_key = 'JDPAY_RSA_PUBLIC_KEY'
```
It's enough.

File: README.md (hunk 4 above)
Comment (genkin-he) [inline README.md:381]: delete redundant lines

File: lib/jd_pay/service.rb (hunk 7 above)
Comment (genkin-he) [inline lib/jd_pay/service.rb]: delete redundant lines

File: lib/jd_pay/service.rb [hunk 10]
```diff
@@ -145,6 +148,17 @@ def notify_verify(xml_str, options = {})
...
       end
 
+      def verify_redirection(params, options = {})
+        params = params.dup
+        sign = params.delete('sign')
+        decrypted_and_dropped = JdPay::Util.decrypt_and_drop_empty(params)
+        decrypted, dropped = decrypted_and_dropped[0], decrypted_and_dropped[1]
+        params = JdPay::Util.stringify_keys(dropped)
+        string = JdPay::Util.params_to_string(params)
+        raise JdPay::Error::InvalidRedirection.new unless Digest::SHA256.hexdigest(string) === JdPay::Sign.rsa_decrypt(sign, options)
```
Comment (genkin-he) [inline lib/jd_pay/service.rb]: It's too long.
use `==`

File: lib/jd_pay/service.rb [hunk 11]
```diff
@@ -175,6 +189,28 @@ def invoke_remote(url, payload, options = {})
...
+        http.use_ssl = true
+        req = Net::HTTP::Post.new(url.to_s)
+        req.form_data = payload
+        req['Content-Type'] = 'application/x-www-form-urlencoded'
+        resp = http.request(req)
+        # This is really a bad design by JDPay.
+        # Examples:
+        # For h5pay: https://h5pay.jd.com/jdpay/payCashier?tradeNum=xxx&orderId=xxx&key=xxx
+        # For webpay: payCashier?tradeNum=xxx&ourTradeNum=xxx&key=xxx
+        if resp['location'].include? base_url
```
Comment (genkin-he) [inline lib/jd_pay/service.rb]:
```ruby
resp['location'].include?(base_url) ? resp['location'] : "#{base_url}/#{resp['location']}"
```

File: lib/jd_pay/util.rb [hunk 12]
```diff
@@ -46,6 +46,34 @@ def build_pay_form(url, form_attributes)
...
+          # Decrypt all valid value
+          if %w(tradeNum amount currency tradeTime note status).include? k
+            v = JdPay::Des.decrypt_3des(v) unless v == ''
+            decrypted[k] = v
+            # According to the doc, any empty value should be dropped.
+            next if v == ''
+            new_hash[k] = v
+          end
+        end
+        return decrypted, new_hash
```
Comment (genkin-he) [inline lib/jd_pay/util.rb]:
```ruby
# Maybe this way is better.
[decrypted, new_hash]
```

File: test/jd_pay/service_test.rb [hunk 13]
```diff
@@ -130,4 +130,116 @@ def test_cancel_user
...
+      tradeName: '25fc4ff3e72f364fc1d80691e88fd7df',
+      amount: 'e5a6c3761ab9ddaf',
+      orderType: 'e00c693e6c5b8a60',
+      notifyUrl: 'd54a8a984359284ec3d58d814e65630300d528555ef1e14ebe9ff87c62d78770',
+      callbackUrl: 'b470f968b7165dcf745e8e129f073434ef9ec43bd64dda9ba2fe77001f4c7347',
+      userId: '29d899fc35dc7aa50640156c06e5ec0f',
+      sign: 'di/QyXcBJGY/Avb8ilTEUsCxto6F5fbjX4mvGt0a0J2XD1dTJ8sPidfTHofagsgz9H/CP3FC6Hqca49woS7o8vkn4oPfA+coHfxBnm4QtBcvdeQjW3fKq2IKbtGG7UGNGXEJc9Gk7BnXNnuvnbEc5FVFcvbqirdNi+7opO9jmaM='
+    }).to_return(headers: {'Location' => 'payCashier?tradeNum=12345678&ourTradeNum=1234567890&key=fookey'}, status: 302)
+
+
```
Comment (genkin-he) [inline test/jd_pay/service_test.rb]: delete redundant lines

File: test/jd_pay/service_test.rb [hunk 14]
```diff
@@ -130,4 +130,116 @@ def test_cancel_user
...
+ U4z14YMka0PDVkzuaA9LV7nPwbQQPj+0tT1F7kiVLIhk//Er2XtzNH93CwReKY9YZEP9NIsgaLTb
+ dK6JAdQiOQyz6CN7dDY=',
+      'status' => 'e00c693e6c5b8a60',
+      'tradeNum' => '38442bada8fcc81581d733ff756dc840272bf690d6b0c0e4',
+      'tradeTime' => 'c33390d9b3d6d6699706ea1467616603a2a1506bbaaf0f59'
+    }
+
+    decrypted_param = JdPay::Service.verify_redirection(params)
+    assert_equal decrypted_param, {"amount"=>"1", "currency"=>"CNY", "note"=>"", "status"=>"0", "tradeNum"=>"dev-1531882041-111", "tradeTime"=>"20180718104826"}
+
```
Comment (genkin-he) [inline test/jd_pay/service_test.rb]: delete redundant lines

**Comment by genkin-he** [comment]
thanks for your work. 
plz:
1.  optimize the code;
2.  could you squash your commits together into one commit with your commit message?

File: lib/jd_pay/service.rb [hunk 15]
```diff
@@ -175,6 +189,28 @@ def invoke_remote(url, payload, options = {})
           }.merge(options)
         )
       end
+
+      def get_redirect_url(url, payload)
+        base_url = WEB_PAY_BASE_URL
+        base_url = H5_PAY_BASE_URL if url === H5_PAY_URL
```
Comment (genkin-he) [inline lib/jd_pay/service.rb]: > === has absolutely nothing whatsoever to do with equality. In particular, it violates pretty much every law that you would ecpect an equality operator to follow. And it does that very much by design

File: README.md (hunk 3 above)
Comment (ZhenhangTung) [reply README.md:276]: For 2:
Nope. If you take a look at https://guides.rubyonrails.org/layouts_and_rendering.html#avoiding-double-render-errors, there should be a return, so the better way could be  `render status: :bad_request and return`

File: lib/jd_pay/result.rb (hunk 6 above)
Comment (ZhenhangTung) [reply lib/jd_pay/result.rb:19]: Oh, that's what you mean. Then it should be named as `decrypt_and_verify `

File: README.md (hunk 9 above)
Comment (ZhenhangTung) [reply README.md:55]: I beleive it's not clear enough? For the new comers, they could misunderstand this `JDPAY_RSA_PUBLIC_KEY ` key. Add comments could help them have a better understanding.

File: lib/jd_pay/util.rb (hunk 8 above)
Comment (ZhenhangTung) [reply lib/jd_pay/util.rb]: Good point. 👍 But if you take a look at the results
```
string = JdPay::Util.params_to_string(params)
        p params.to_param
        p string
```
Results:
```
"amount=1&currency=CNY&note=test+product&status=0&tradeNum=dev-1531992606-125&tradeTime=20180719173221"
"amount=1&currency=CNY&note=test product&status=0&tradeNum=dev-1531992606-125&tradeTime=20180719173221"
```
I beleive `to_query` will endode the string, which will lead to an error on the verification

**Comment by ZhenhangTung** [comment]
About squashing commits, please take a look at https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits. I believe we don't need to do it manually.

File: lib/jd_pay/util.rb (hunk 8 above)
Comment (genkin-he) [reply lib/jd_pay/util.rb]: you misunderstood me on this.
rename the `to_uri` using the `to_query`, there is a method named `to_uri`.

File: README.md (hunk 3 above)
Comment (genkin-he) [reply README.md:276]: this way to early return from a rails controller. It's unnecessary at the end of a action.

**Comment by genkin-he** [comment]
Special thanks to you

**Pull request merged**